gnucash/gnome-utils/CMakeLists.txt

```diff
@@ -264,9 +264,9 @@ gnc_add_scheme_targets(scm-gnome-utils-1
 
 gnc_add_scheme_targets(scm-gnome-utils-2
   gnc-menu-extensions.scm
-  ""
+  "gnucash/gnome-utils/include"
```

Comment:
I'm not married to the name itself -- it was the best I could come up with to indicate that these files are not guile modules themselves but files that will be included in guile modules via "load" or "primitive-load-from-path".

I started this reorganisation because I continued to get confused  by what our scheme api really consists of. I wanted to get to something that was clear, consistent and logical.

So first I set out to isolate the module files (that is all files starting with `(define-module (xyz))`). Those represent our SCM api and are the only interface we want to expose (to the rest of gnucash or even external callers).

There were plenty of other scm files splattered about in the scm directory. Looking at these I found they are all only included in exactly one module file. So I started to look at a "module" as the whole of the scm module file with the files included in it.  For example the "engine" module consists of `engine.scm` and 4 files it includes: `commodity-table.scm`,  `engine-interface.scm`,  `engine-utilities.scm` and  `gnc-numeric.scm`. They could all have been in on huge single `engine.scm` file, but that would be unwieldy. However these 4 files have no meaning outside of engine either.

So to tidy this up, I want to move these non-module files with the modules they belong to. I borrowed the python module structure of having a file `module.scm` and everything which belongs to that module in a subdirectory called `module`

The `include` subdirectory came from the idea that I want to stress these files are *not* modules in themselves. For all but the report module this comes with a seemingly unneeded indirection.

However for the report it's a bit more involved: that module has submodules, like reports, stylesheets,... It makes sense to me these submodules are stored in the report directory. If I *don't* add an include directory for the ordinary non-module scm files that would mean the report directory would now have a mix of module files and non-module files and the confusion would continue one level down the module structure. That's why I came up with the include directory.

For consistency I decided to add it to all other modules as well.

However I'm open to suggestions for improvement.
For example I could remove the include from the path for all but the report module. Would you like that more ?
In report I would like to continue to separate these module private files from the other module files. I could use another name though. How about `private` ?

Reply:
@gjanssens that's not quite how Python modules work, but it doesn't matter. Your basic scheme of module_name.scm and a module_name subdirectory containing the module's implementation is fine, though from the source side the fact that most of the module_name.scm are created by SWIG from C sources makes the result a bit less clear, and I think that's the basic flaw in the plan: While the very few users writing custom reports might look at the installation directory structure to figure out the Scheme API, the rest of us live in the source tree and reconfiguring the installation tree doesn't do much good.

> For example I could remove the include from the path for all but the report module. Would you like that more ?

Yes

> In report I would like to continue to separate these module private files from the other module files. I could use another name though. How about private ?

The only not-modules in ```report``` are the three eguile and css files that belong to the eguile reports and and 18 of the 24 scm files in report-system. ISTM the 6 module files in report-system should be moved to the ```report``` directory and their respective dependencies moved to like-named subdirectories, just like engine, app-utils, and the rest. Going a step further, to simplify the ```use-modules``` usage we could move all of the report-system modules to ```share/gnucash/scm/gnucash``` and put only reports in ```share/gnucash/scm/gnucash/reports```.  The eguile reports can each have a like-named (e.g. ```balsheet-eg```) subdirectory for their eguile and css files. The TXF reports will need a little more massaging as their infrastructure is a bit more spread out.

Reply:
> @gjanssens that's not quite how Python modules work, but it doesn't matter.
Yeah, I'm mixing things up here...

> Your basic scheme of module_name.scm and a module_name subdirectory containing the module's implementation is fine, though from the source side the fact that most of the module_name.scm are created by SWIG from C sources makes the result a bit less clear,

I feel relieved I'm not the only one that frequently gets confused by this complex interaction between our C and scheme code :)
SWIG does not generate scm files at all. It generates a C file that is compiled and linked into our *.so files to expose C functions to guile code. This is further obfuscated because we have added the gnc-module abstraction layer for automatic module initialization. Because it's almost impossible to track which part really depends on which other we add module loading and *.so file loading in almost all guile modules and guile module loading in almost every gnc-module. Another layer of complexity to untangle.

> and I think that's the basic flaw in the plan: While the very few users writing custom reports might look at the installation directory structure to figure out the Scheme API, the rest of us live in the source tree and reconfiguring the installation tree doesn't do much good.

The whole mixed language stuff affects many parts in both source and install hierarchies. I agree the source side could use a similar cleanup. But I had to start somewhere and restructuring the install hierarchy has at least helped me to get the picture a little clearer. As time permits I plan to continue poking at this and clear it out even more. My long term vision is to move all foreign language bits to the bindings directory (we're not there yet by a long shot). And eliminate the gnc-module abstraction layer from the binding mix. But those are not the topics of this PR so I propose to discuss that in due time.
> 
> > For example I could remove the include from the path for all but the report module. Would you like that more ?
> 
> Yes
> 
Ok, will do.

> > In report I would like to continue to separate these module private files from the other module files. I could use another name though. How about private ?
> 
> The only not-modules in `report` are the eguile and css files that belong to the three eguile reports and and 18 of the 24 scm files in report-system. ISTM the 6 module files in report-system should be moved to the `report` directory and their respective dependencies moved to like-named subdirectories, just like engine, app-utils, and the rest.

I went a bit further and moved `report-system.scm` up to the same level as `engine.scm` as that's really the public interface of the whole report system. The other 5 modules only make sense if the report-system itself is setup, hence I kept them as submodules.
Note I did rename `report-system.scm` to `report.scm` to go with the `report` directory instead of the other way around. If I had chosen to rename the directory to report-system instead that would have required all reports to be defined as `(define-module (gnucash report-system reports <rptname>))` rather than `(define-module (gnucash report reports <rptname>))`. The latter is a tad shorter.
Also note this prompted me to rename the already existing non-module `report.scm` to `report-core.scm` to avoid more confusion.
The 18 non-module files originally in `report-system` are all loaded in `report-system.scm` (now `report.scm`) so logically they should be stored in the `report` directory. However that also has the 5 report submodule scm files which I want to keep separate from the non-module files.
> Going a step further, to simplify the `use-modules` usage we could move all of the report-system modules to `share/gnucash/scm/gnucash` and put only reports in `share/gnucash/scm/gnucash/reports`.

I had considered this but in the end preferred keeping all report related things together as the 5 submodules don't make sense outside of the report context.
It does mean I need to find another way to separate the non-module scm files from the others.
> The eguile reports can each have a like-named (e.g. `balsheet-eg`) subdirectory for their eguile and css files. 

I do like this idea.
>The TXF reports will need a little more massaging as their infrastructure is a bit more spread out.

Indeed.

Reply:
> I went a bit further and moved report-system.scm up to the same level as engine.scm as that's really the public interface of the whole report system. The other 5 modules only make sense if the report-system itself is setup, hence I kept them as submodules.
Note I did rename report-system.scm to report.scm to go with the report directory instead of the other way around. If I had chosen to rename the directory to report-system instead that would have required all reports to be defined as (define-module (gnucash report-system reports <rptname>)) rather than (define-module (gnucash report reports <rptname>)). The latter is a tad shorter.
Also note this prompted me to rename the already existing non-module report.scm to report-core.scm to avoid more confusion.
The 18 non-module files originally in report-system are all loaded in report-system.scm (now report.scm) so logically they should be stored in the report directory. However that also has the 5 report submodule scm files which I want to keep separate from the non-module files.

I think I wasn't clear enough, so here's the layout I'm proposing:
```
    gnucash/scm/gnucash/report/:
           collectors.scm
           eguile-gnc.scm 
           eguile-html-utilities.scm
           eguile-utilities.scm
           report-collectors.scm
           report-system.scm  ;; could be just 'system' so one would
           report-system/         ;; (use-modules (gnucash report system))
                commodity-utilities.scm
                html-acct-table.scm
                ...
                html-utilities.scm
                options-utilities.scm
                report-utilities.scm
                report-core.scm
                trep-engine.scm
         reports/
               account-piecharts.scm
               ...
               balsheet-eg.scm
               balsheet-eg/
                    balsheet-eg.eguile.scm
                    balsheet-eg.css
              etc.
         stylesheets.scm ;;  This is sort-of cheating because each stylesheet is a module and
         stylesheets/        ;;   stylesheets.scm isn't, but it helps declutter the root directory a bit.
              stylesheet-easy.scm
              stylesheet-fancy.scm
              stylesheet-footer.scm
              stylesheet-head-or-tail.scm
              stylesheet-plain.scm
```

Reply:
Isn't @christopherlam trying to get rid of ```collectors.scm``` and ```report-collectors.scm```?

Reply:
> Isn't @christopherlam trying to get rid of `collectors.scm` and `report-collectors.scm`?

They are indeed removed in the guile deprecation cleanup commit of this branch.

Reply:
> > Isn't @christopherlam trying to get rid of `collectors.scm` and `report-collectors.scm`?
> 
> They are indeed removed in the guile deprecation cleanup commit of this branch.

Hmm, they should have been in that commit, but I accidentally did it in the commit folding report-system into report. Fixed that now and pushed separately to master.

Reply:
As for your layout, I did understand your proposal.
I'll use `report-system` to explain rather than `report` in both proposals to have one less moving target.
The difference between your suggestion and mine boils down to at what level we want to access the report system:
Your proposal puts the report system in a subdirectory of report `(gnucash report report-system)`.
Mine puts the report system at the same level as engine.scm and most other modules `(gnucash report-system)`.

Yours has the advantage you automatically get the subdirectory for the report-system included files. Mine gives a more consistent public view of the modules:
* want to use engine features: `(use-modules (gnucash engine))`
* want to use report system features: `(use-modules (gnucash report-system))`

Your proposal doesn't have a report module at that gnucash level, you'll need to know report is slightly different and would require `(use-modules (gnucash report report-system))`. That's the minor inconsistency I tried to solve still at the cost of an extra subdirectory to add the load files.

While I prefer a consistent public module view, I don't feel that strongly about it. So if you really don't want it I'm fine using your structure as well.

Reply:
An observation on the stylesheets. As you hint at already stylesheets.scm is only a shim to load the other stylesheets. I would propose to reuse the report loader to also load the stylesheets. I can do this by moving all stylesheets to report/reports/stylesheets and add "stylesheets" as extra subdirectory in
https://github.com/Gnucash/gnucash/blob/master/gnucash/report/reports/reports.scm#L46

The libgncmod-stylesheets.so module also does almost nothing. Aside from loading the stylesheets (which I can delegate to the report loader) it adds one menu item via gnc-plugin-stylesheets. I propose to move this plugin to gnome instead and eliminate the libgncmod-stylesheets.so module. One module less to worry about.

Reply:
Hmm, so what's left in the report(-system) directory afterwards ?
* jqplot is on its way out. One final report (currently not included in the build) to migrate and that directory can be removed
* locale-specific - a closer look at the code suggests this directory is essentially redundant. The two scm files in here are still loaded, but they are really shims to load the tax reports themselves. These reports are again loaded in from c code so the shims could go away.
* the eguile based report support files (for balsheet, receipt and tax-invoice) - I'm not sure it will be feasible to store them in subdirectories relative to their base report. Or more precisely I'm not sure it will be easy to load them from such fairly arbitrary locations. Our current load mechanism starts from a fixed set of paths that doesn't involve report names. Regardless they are not part of the core report system. Instead they are support files for specific reports so they should find a home somewhere under the reports directory as well. The css files are easy: they are stylesheets and we will have a stylesheets directory there if I finish the stylesheet move proposed above. The eguile template files may end up in a templates directory as a precursor to a future template based report system.
* two locale tax related files taxtxf.scm and taxtxf-de_DE.scm - these are the actual locale specific reports. Installing them in reports/locale-specific instead would be better.

Moving all of that around makes for a rather small report "module": a reports subdirectory with three eguile modules and 20 plain scm files.

I'm ready to revisit my point of view on how this should be modelled. I now propose:
```
gnucash/scm/gnucash/
    report-system.scm
    report-system/
        eguile-gnc.scm 
        eguile-html-utilities.scm
        eguile-utilities.scm
        internal/
            commodity-utilities.scm
            html-acct-table.scm
            ...
            html-utilities.scm
            options-utilities.scm
            report-utilities.scm
            report-core.scm
            trep-engine.scm
    reports/
        example/
            average-balance.scm
            daily.scm
            ...
        locale-specific/
            taxtxf-de_DE.scm
            taxtxf.scm
        standard/
            account-piecharts.scm
            ...
            balsheet-eg.scm
            ...
        stylesheets/
            balsheet-eg.css
            receipt-eg.css
            stylesheet-easy.scm
            stylesheet-fancy.scm
            stylesheet-footer.scm
            stylesheet-head-or-tail.scm
            stylesheet-plain.scm
            taxinvoice-eg.css
        templates/
            balsheet-eg.eguile.scm
            receipt-eg.eguile.scm
            taxinvoice-eg.eguile.scm
```
The major changes here are
* promote the `reports` module to become a top-level gnucash module (one level up). This module now has the most substructure and I can equally see it stand on its own.
* Squash what remains of the report directory together with the report-system directory and promote this report-system module to become a top-level module as well. In the end it may be better to keep `report-system` as name for that top-level module rather than my previously proposed `report` to make it more distinct from reports.
* The clash of modules and loadables would still exist in the report-system module so I continue to propose putting the loadables in a subdirectory to ease at-a-glance module recognition. I'll offer a different name: `internal` ?

Reply:
> Your proposal doesn't have a report module at that gnucash level, you'll need to know report is slightly different and would require (use-modules (gnucash report report-system)). That's the minor inconsistency I tried to solve still at the cost of an extra subdirectory to add the load files.

Because I (mis?)understood that you wanted to keep all of the report stuff under ```report/```.

That inconsistency still exists with your new layout: The eguile and stylesheets modules still have extra steps in their use-modules paths. One could dismiss that as simply a reflection of the much greater complexity of the report system, bearing in mind that the main reason for all of the other scheme API is to support reports; even options and fin are the way they are to make them accessible from reports. OTOH one could also transform eguile into a similar layout, making eguile-gnc.scm, eguile-html-utilities.scm, and eguile-utilities.scm not-modules and having a shim module to load-from-path those three and export their public functions. 

It looks to me like there's no need for ```stylesheet-foo.scm``` to be a module: They don't export API in the usual way; instead they call ```gnc:define-html-style-sheet``` with their otherwise private functions, and ```html-style-sheets.scm``` uses those functions to style the report. I think that means that they belong with report-system rather than reports.

+10 to removing gnc-modules from report*, it just adds to the confusion and make it harder to find where API is used.

Reply:
Coming up with a solid design is hard as the way to divide depends on what is we consider most important.

> Because I (mis?)understood that you wanted to keep all of the report stuff under `report/`.
> 

Ok, I see now my descriptions were a bit fuzzy. When trying to come up with an improved api I started from the module view an api consumer would see. That is if some code needed scheme access to engine features they should be able to do `(use-modules (gnucash engine))`, that is, load the engine odule from the gnucash project. For the smaller modules that is easily implemented. The only blob not adhering to this now the the whole report system. It's also a much more elaborate subsystem. And either we flatten it out (partly or completely) or remodel its substructure. I can't make up my mind which one would be best in the end.

The original idea of keeping all report related code together is based on the way our source tree is structured and (for now) how we have a similar module hierarchy for our C code. That is we have libgncmod-engine.so, libgncmod-app-utils.so and libgncmod-report.so and so on. So to reuse muscle memory it is helpful to stick to that same structure for the guile modules: have an engine.scm, app-utils.scm, report.scm, resulting in api calls like `(use-modules (gnucash engine))`, `(use-modules (gnucash app-utils))` or `(use-modules (gnucash report))`. This latter does require the report.scm file be at the same level of engine.scm in the installation path and this is the part I took for granted implicitly.
In the source structure the other report related modules like the reports, stylesheets and so on are all in subdirectories of the report directory. Again to reuse the mental model we already have from there I originally thought it better to keep all these other related modules as submodules of the report one, so the api for those would become `(use-modules (gnucash report reports))` or `(use-modules (gnucash report eguile-gnc))`.

> That inconsistency still exists with your new layout: The eguile and stylesheets modules still have extra steps in their use-modules paths. One could dismiss that as simply a reflection of the much greater complexity of the report system, bearing in mind that the main reason for all of the other scheme API is to support reports; even options and fin are the way they are to make them accessible from reports. OTOH one could also transform eguile into a similar layout, making eguile-gnc.scm, eguile-html-utilities.scm, and eguile-utilities.scm not-modules and having a shim module to load-from-path those three and export their public functions.
> 

Thinking this through that's probably the best path at this point. It's likely eguile will eventually be replaced with a more mature and better maintained templating system.

> It looks to me like there's no need for `stylesheet-foo.scm` to be a module: They don't export API in the usual way; instead they call `gnc:define-html-style-sheet` with their otherwise private functions, and `html-style-sheets.scm` uses those functions to style the report. I think that means that they belong with report-system rather than reports.

Yes and no. By that same reasoning none of the reports need to be modules. They don't export API either and instead they call `gnc:define-report` with their otherwise private functions and and `report.scm` (or `report-core.scm` after restructuring) uses those functions to expose the reports to the user interface. On the other hand encapsulating each report or stylesheet in a module adds namespace isolation. That is each report can have the exact same (private) variable name and function definitions like `reportname` or `options-generator`. That greatly helps for custom-report builders. Strictly speaking this could have been done as well for the renderer function but for reasons unknown to me those are uniquely defined.

In all I don't intend take on this aspect as well in this round of cleanups. Instead I would like to treat reports and stylesheets in the same way, that is load them all automatically when found in a pre-defined directory.

So hierarchy so far: like my previous comment, with two changes:
the eguile stuff will be taken out of the report-system directory into its own top-level module and the loadables in report-system go directly in report-system. Is that reasonable ?

Reply:
>So hierarchy so far: like my previous comment, with two changes:
the eguile stuff will be taken out of the report-system directory into its own top-level module and the loadables in report-system go directly in report-system. Is that reasonable ?

Yes, that's what I've been arguing for.

> In all I don't intend take on this aspect as well in this round of cleanups. Instead I would like to treat reports and stylesheets in the same way, that is load them all automatically when found in a pre-defined directory.

Fair enough. That said,

> On the other hand encapsulating each report or stylesheet in a module adds namespace isolation. That is each report can have the exact same (private) variable name and function definitions like reportname or options-generator. 

That can be achieved by (in the case of reports) moving the ```(gnc:define-report``` line to the top of the the file so that the subsequent function definitions are inside its scope. 

You skipped over that stylesheets are used by report-system, not by individual reports (though they are used by individual report tests, though I don't understand yet how and why), and so should go into ```report-system/``` rather than ```reports/```.

Reply:
> That can be achieved by (in the case of reports) moving the (gnc:define-report line to the top of the the file so that the subsequent function definitions are inside its scope.

No, on reflection that's obviously wrong. I mean to use a closure instead of a module to limit the scope of the function variables. One way would be to replace the defined functions with lambdas in the ```gnc:define-report``` call, but that would be pretty ugly.

Reply:
> You skipped over that stylesheets are used by report-system, not by individual reports (though they are used by individual report tests, though I don't understand yet how and why), and so should go into `report-system/` rather than `reports/`.

Right, so that brings us to a practical trade-off decision. Currently the report-loading code is in reports.scm. To reuse that code for stylesheets, I see these options:
* bring stylesheets under reports. That's the option I had chosen as that could also serve as the location for the eguile css stylesheets which are directly used by the eguile reports. Eguile may not be the future but unless we completely abandon html, I do think css stylesheets will be.
* keep stylesheets under report-system and more or less duplicate reports.scm's report loading code into a similar `report-system/stylesheets.scm`. I don't think we want this as that would resurface the conflict between loadables and modules in that directory.
* move the code report loading code into the already very large report-system.scm to be able to load from both `report-system/stylesheets` and `reports/{example,standard}`.  We could but then I would argue to move the whole reports thing back under report as well as there would be no relevant code left to justify a separate reports module. Obviously the reports themselves would remain as modules for now.

Reply:
Ah, that naked ```for-each``` in reports.scm was the piece I was missing. Do I understand correctly that that's run as a side-effect of ```(use-modules (gnucash reports))```? If so, yuck.

Reply:
Call me dense, but why "yuck" ?

Reply:
For completeness, we resolved this [in IRC](https://wiki.gnucash.org/logs/2019/06/13.html#T14:09:08).